packages/edit-navigation/src/store/actions.js

```
@@ -49,171 +39,380 @@ export function setSelectedMenuId( menuId ) {
  * @param {Object} post A navigation post to process
  * @return {Function} An action creator
  */
-export const createMissingMenuItems = serializeProcessing( function* ( post ) {
+export const createMissingMenuItems = ( post ) => async ( {
+	dispatch,
+	registry,
+} ) => {
 	const menuId = post.meta.menuId;
+	// @TODO: extract locks to a separate package?
+	const lock = await registry
+		.dispatch( coreDataStore )
+		.__unstableAcquireStoreLock( STORE_NAME, [ 'savingMenu' ], {
+			exclusive: false,
+		} );
+	try {
+		// Ensure all the menu items are available before we start creating placeholders.
+		await registry
+			.resolveSelect( coreDataStore )
+			.getMenuItems( { menus: menuId, per_page: -1 } );
 
-	const mapping = yield getMenuItemToClientIdMapping( post.id );
-	const clientIdToMenuId = invert( mapping );
-
-	const stack = [ post.blocks[ 0 ] ];
-	while ( stack.length ) {
-		const block = stack.pop();
-		if ( ! ( block.clientId in clientIdToMenuId ) ) {
-			const menuItem = yield apiFetch( {
-				path: `/__experimental/menu-items`,
-				method: 'POST',
-				data: {
-					title: 'Placeholder',
-					url: 'Placeholder',
-					menu_order: 1,
-				},
-			} );
+		const menuItemIdToBlockId = await dispatch(
+			getEntityRecordIdToBlockIdMapping( post.id )
+		);
+		const knownBlockIds = new Set( Object.values( menuItemIdToBlockId ) );
 
-			mapping[ menuItem.id ] = block.clientId;
-			const menuItems = yield resolveMenuItems( menuId );
-			yield dispatch(
-				'core',
-				'receiveEntityRecords',
-				'root',
-				'menuItem',
-				[ ...menuItems, menuItem ],
-				menuItemsQuery( menuId ),
-				false
-			);
+		const blocks = blocksTreeToFlatList( post.blocks[ 0 ].innerBlocks );
+		for ( const { block } of blocks ) {
+			if ( block.name !== 'core/navigation-link' ) {
+				continue;
```

Maybe fail loudly here? In case a plugin has added a non-link block or something weird like that.

---

There are blocks we want to skip, e.g. a submenu block. Logging a warning seems to be a good trade-off though.

---

Do we not hide submenu if it's not supported?

---

Submenu blocks should be converted to menu items as well btw.

---

Tracked in https://github.com/WordPress/gutenberg/issues/35045